service/controller/v12/resource/hostedzone/resource.go

```diff
@@ -0,0 +1,154 @@
+package hostedzone
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/service/route53"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+
+	controllercontext "github.com/giantswarm/aws-operator/service/controller/v12/context"
+	"github.com/giantswarm/aws-operator/service/controller/v12/key"
+)
+
+const (
+	name = "hostedzonev12"
+)
+
+type Config struct {
+	HostRoute53 *route53.Route53
+	Logger      micrologger.Logger
+}
+
+type Resource struct {
+	hostRoute53 *route53.Route53
+	logger      micrologger.Logger
+}
+
+func New(config Config) (*Resource, error) {
+	if config.HostRoute53 == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.HostRoute53 must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	r := &Resource{
+		hostRoute53: config.HostRoute53,
+		logger:      config.Logger,
+	}
+
+	return r, nil
+}
+
+func (r *Resource) Name() string {
+	return name
+}
+
+func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+	err := r.setStatus(ctx, obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+func (r *Resource) EnsureDeleted(ctx context.Context, obj interface{}) error {
+	err := r.setStatus(ctx, obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+// setStatus searches for HostedZone in AWS API by name for their IDs. Those
+// IDs are set in controller context status for further use.
+func (r *Resource) setStatus(ctx context.Context, obj interface{}) error {
+	customObject, err := key.ToCustomObject(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	controllerCtx, err := controllercontext.FromContext(ctx)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", "looking for host cluster hosted zone IDs in AWS API")
+
+	var (
+		apiFound     = false
+		etcdFound    = false
+		ingressFound = false
+
+		apiDomain     = key.HostedZoneNameAPI(customObject)
+		etcdDomain    = key.HostedZoneNameEtcd(customObject)
+		ingressDomain = key.HostedZoneNameIngress(customObject)
+	)
+
+	var marker *string
+	for {
+		in := &route53.ListHostedZonesInput{
+			Marker: marker,
+		}
+
+		out, err := r.hostRoute53.ListHostedZones(in)
```

Comment:
Does ListHostedZones return all zones or is there paging?

I got hit by this with the loadbalancer resource where the customer had a lot of ELBs in their account and API only returns 20.

Reply:
Nice, good way to handle the paging.